refactor error page composition

jroebu14 opened the pull request:
Resolves to https://github.com/bbc/simorgh/issues/5482

**Overall change:** 
Moves `ErrorMain` into the `pages` directory (to align with other pages) and applies page enhancing HOCs in `pages/index.js` where eventually all pages will have their enhancers applied.

**Code changes:**
- Moves `src/app/containers/ErrorMain/index.jsx` (along with Jest tests and snapshots) to `src/app/pages/ErrorPage/index.jsx` so that the `pages` directory contains the page components.
- Applies page enhancing HOCs in `pages/index.js` that will eventually apply all HOCs to pages and export them.
- Deletes `src/app/containers/ErrorMain` directory since it was moved into the `pages` directory.
- Renames `ErrorPage` in the `components` directory to `ErrorMain` to avoid confusion between it and the actual error page component. I chose the name `ErrorMain` since it actually contains the `main` html element.
---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [ ] This PR requires manual testing

Commit:
refactor error page composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

ryanmccombe reviewed (approved):
👍

chris-hinds reviewed (approved):
👍

Commit:
fix translation getting logic in ErrorPage

sareh reviewed (changes requested):
This seems to break rendering 404 error pages.

Visiting this 404 error page, I see a 500 page outputted http://localhost:7080/igbo/404
<img width="1022" alt="Screenshot 2020-02-28 at 18 15 54" src="https://user-images.githubusercontent.com/3028997/75574563-6e27a280-5a56-11ea-8d75-aa0cb3ae767b.png">
The relevant cypress tests are failing for this reason too. 
```
errorPage404 - igbo - Canonical
       Tests for igbo errorPage404
         igbo error page routes
           /igbo/404 should have response code 200:
            CypressError: Timed out retrying: expected '<span.StatusCode-sc-1h0f93q-0.fOcHxq>' to contain '404
```

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

Commit:
add more tests for error pages

Commit:
refactor route names

Commit:
Merge branch 'latest' into refactor-error-page-composition

jroebu14 commented:
@sareh this happened because I didn't correctly apply [this error logic](https://github.com/bbc/simorgh/pull/5714/files#diff-3190e4bcf10d8e048952c0dbe9c6050fL11) which I have now [fixed](https://github.com/bbc/simorgh/pull/5714/commits/3adccfaab5a39a10ba09df14dae790253800036c)

It's highlighted that there are no unit tests for these routes so I wrote some [here](https://github.com/bbc/simorgh/pull/5714/commits/a88d35a58249957c4aae126abc7f21cae2753141)

I also decided to rename the error routes because I just realised that `error200` sounds weird so I renamed `error200` to `error` which matches those `igbo/500` and `igbo/404` pages and I renamed `error404` to `404`.

Commit:
remove default errorCode 500 from component

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

sareh commented on src/app/routes/index.test.jsx:
```diff
@@ -222,7 +246,23 @@ it('should route fallback to and render a 500 error page', async () => {
 
-it('should fallback to and render a 404 error page', async () => {
+it('should route to and render a 404 error page', async () => {
+  const pathname = '/igbo/404';
```
Thank you for adding these unit tests.

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

sareh committed:
Merge branch 'latest' into refactor-error-page-composition

sareh commented on src/app/pages/ErrorPage/index.jsx:
```diff
@@ -2,7 +2,7 @@ import React, { useContext } from 'react';
 import Helmet from 'react-helmet';
 import { ServiceContext } from '#contexts/ServiceContext';
-import ErrorPageComponent from '../../components/ErrorPage';
+import ErrorMain from '../../components/ErrorMain';
```
```suggestion
import ErrorMain from '#app/components/ErrorMain';
```

sareh commented on src/app/pages/ErrorPage/index.jsx:
```diff
@@ -33,7 +33,7 @@ const ErrorMetadata = ({ dir, lang, messaging, brandName, themeColor }) => {
 };
 
-const ErrorMain = ({ status }) => {
+const ErrorPage = ({ status, errorCode }) => {
```
Can we update this component to only take in `errorCode` as a prop. We'd just need to update the `ErrorPage` prop name in `PageHandlers/withData/index.jsx` and `PageHandlers/withError/index.jsx` but it'd simplify this component even further.

sareh commented on src/app/routes/index.js:
```diff
@@ -2,7 +2,7 @@ import article from './article';
-import error200 from './error200';
-import error404 from './error404';
+import error from './error';
+import notFound404 from './404';
```
Can you please rename the 404 page as errorNoRouteMatch. It ensures it's colocated with the error page and it has a distinction between the purpose of this one as opposed to `error`.
```suggestion
import notFound404 from './errorNoRouteMatch';
```

sareh reviewed (commented):
Thanks for the changes, Jonathan. This looks much better. I've added some comments below.

Commit:
Update src/app/pages/ErrorPage/index.jsx

Co-Authored-By: Sareh <sareh@users.noreply.github.com>

Commit:
refactor error page props. improve error page tests

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
add unit test

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

sareh commented on src/app/routes/error/index.js:
```diff
@@ -5,6 +5,11 @@ export default {
   component: ErrorPage,
-  getInitialData: () => Promise.resolve({ status: 200 }),
+  getInitialData: pathname => {
+    const [, , errorPage] = pathname.split('/');
```
This array destructoring looks as though it's a mistake, instead of a deliberate task of getting the string '404'/'500' from the pathname. 
Is it possible to get the error code string via React router methods instead, rather than us extracting it manually here? This would make this code less brittle, so that if the regex updates, this line won't break the Error page implementation.

jroebu14 replied:
You're right this is much better https://github.com/bbc/simorgh/pull/5714/commits/d008b252514689501a78501f65cf64279311dedd

sareh commented on src/app/pages/ErrorPage/index.jsx:
```diff
@@ -54,18 +54,17 @@ const ErrorMain = ({ status }) => {
-ErrorMain.propTypes = {
-  status: number.isRequired,
+ErrorPage.propTypes = {
+  errorCode: number,
```
I think this prop should be required. Since we never want to render this page without the errorCode having a value.

jroebu14 replied:
There is a fallback to a 500 if we don't provide one but I've made it required anyway 
https://github.com/bbc/simorgh/pull/5714/commits/ca02bbfe24378e97aeb1582147188f99fc1927e9

Commit:
Merge branch 'latest' into refactor-error-page-composition

sareh reviewed (commented):
Thanks for the update. It looks good. Just a couple of comments.

Commit:
update prop types

Commit:
use react router to get params

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
matchPath will never be null here

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
fix test

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

sareh reviewed (approved):
👍 Apologies for the delay. This looks great. 
<img src="https://media.giphy.com/media/xUPGcxpCV81ebKh7Vu/giphy.gif" alt="thanks a bunch" />

karinathomasbbc committed:
Merge branch 'latest' into refactor-error-page-composition

Commit:
if errorCode does not come from route params, use errorCode in initialData

Commit:
Merge branch 'refactor-error-page-composition' of github.com:bbc/simorgh into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition

Commit:
Merge branch 'latest' into refactor-error-page-composition
